core/src/Revolution/Sources/modMediaSource.php

```diff
@@ -354,41 +358,44 @@ public function getContainerList($path)
         $allowedExtensions = $this->getAllowedExtensionsArray($properties);
 
         $directories = $dirNames = $files = $fileNames = [];
+
         if (!empty($path)) {
+
+            // Ensure the provided path can be read.
             try {
-                $meta = $this->filesystem->getMetadata($path);
-            } catch (Exception $e) {
+                $mimeType = $this->filesystem->mimeType($path);
+            } catch (FilesystemException | UnableToRetrieveMetadata $e) {
                 $this->addError('path', $e->getMessage());
-
+                $this->xpdo->log(modX::LOG_LEVEL_ERROR, $e->getMessage());
                 return [];
             }
-            if (isset($meta['type']) && $meta['type'] != 'dir') {
-                $this->addError('path', $this->xpdo->lexicon('file_folder_err_invalid'));
 
+            if ($mimeType !== 'directory') {
+                $this->addError('path', $this->xpdo->lexicon('file_folder_err_invalid'));
                 return [];
             }
+
         }
 
         try {
             $re = '#^(.*?/|)(' . implode('|', array_map('preg_quote', $skipFiles)) . ')/?$#';
-            /** @var array $contents */
             $contents = $this->filesystem->listContents($path);
```

Comment:
Can we refactor the code below to implement the ["filter" functionality](https://flysystem.thephpleague.com/v2/docs/usage/directory-listings/)?

Reply:
Hmm.. do you mean filtering out the $skipFiles array? e.g.
```
Array
(
    [0] => .svn
    [1] => .git
    [2] => _notes
    [3] => nbproject
    [4] => .idea
    [5] => .DS_Store
    [6] => .
    [7] => ..
)

```
`StorageAttributes` doesn't appear to be able to get the file extension (at least without taking it from the end of the `path`)
I'm not seeing any real benefit to refactoring it... unless you mean something else?

Reply:
Instead of the various if statements with a continue I was thinking about something like this:
```
$contents = $this->filesystem->listContents($path)
->filter(function(StorageAttributes $attributes) use ($re) {
    return !preg_match($re, $attributes->path());
})
->filter(function(StorageAttributes $attributes) use ($properties) {
    if ($attributes->isDir()) {
        return $this->hasPermission('directory_list');
    } elseif ($attributes->isFile()) {
        return $this->hasPermission('file_list') && !$properties['hideFiles'];
    }

    return false;
})
;
```

_Note: code might not actually work, it's just a rough sketch of my idea_

Reply:
Ah very cool!